programs/ziti-edge-tunnel/instance-config.c

```diff
@@ -0,0 +1,165 @@
+    }
+    return mutex_init;
+}
+
+bool load_config_from_file(char* config_file_name) {
+    bool loaded = false;
+
+    FILE* config_file = fopen(config_file_name, "r");
+    if (config_file != NULL) {
+        char *config_buffer = malloc(MAX_BUFFER_LEN * sizeof(char));
+        *config_buffer = NULL;
+        char line[512];
+        while ((fgets(line, sizeof(line), config_file)) != NULL) {
+            strcat(config_buffer, line);
+        }
+
+        if (strlen(config_buffer) > 0) {
+            loaded = load_tunnel_status(config_buffer);
+            if (!loaded) {
+                ZITI_LOG(WARN, "Config file %s cannot be read, will be overwritten", config_file_name);
+            }
+        }
+        fclose(config_file);
+    } else {
+        ZITI_LOG(INFO, "The config file %s does not exist. This is normal if this is a new install or if the config file was removed manually", config_file_name);
+    }
+    return loaded;
+}
+
+bool load_tunnel_status_from_file(uv_loop_t* ziti_loop) {
+    char* config_path = get_system_config_path();
+
+    uv_fs_t fs;
+    int check = uv_fs_mkdir(ziti_loop, &fs, config_path, 0755, NULL);
+    if (check == 0) {
+        ZITI_LOG(TRACE,"config path is created at %s", config_path);
+    } else {
+        ZITI_LOG(TRACE,"config path is found at %s", config_path);
+    }
+    bool loaded = false;
+
+    char* config_file_name = get_config_file_name(config_path);
+    char* bkp_config_file_name = get_backup_config_file_name(config_path);
+
+    // try to load tunnel status from config file
+    loaded = load_config_from_file(config_file_name);
+
+    // try to load tunnel status from backup config file
+    if (!loaded) {
+        loaded = load_config_from_file(bkp_config_file_name);
+    }
+
+    // not able to load the tunnel status from both the config and backup files
+    if (!loaded) {
+        ZITI_LOG(WARN, "Config files %s and the backup file cannot be read or they do not exist, will create a new config file or the old one will be overwritten", config_file_name);
+    }
+
+    free(config_file_name);
+    free(bkp_config_file_name);
+    free(config_path);
+    return loaded;
+}
+
+bool save_tunnel_status_to_file() {
+    tunnel_status* stat = get_tunnel_status();
+    size_t json_len;
+    char* tunnel_status = tunnel_status_to_json(stat, 0, &json_len);
+    bool saved = false;
+
+    if (json_len > 0) {
+        char* config_path = get_system_config_path();
+
+        char* config_file_name = get_config_file_name(config_path);
+        char* bkp_config_file_name = get_backup_config_file_name(config_path);
+
+        int stat = uv_mutex_trylock(&mutex);
+        if (stat == 0) {
+            //copy config to backup file
+            FILE* config = fopen(config_file_name, "r");
+            if (config == NULL) {
+                ZITI_LOG(ERROR, "Could not open config file %s", config_file_name);
+            } else {
+                FILE* backup_config = fopen(bkp_config_file_name, "w");
```

Comment:
why not just name the file?

Reply:
changed the code, used rename instead of file overwrite operation